### PR TITLE
Dev multi optimizer

### PR DESCRIPTION
**Pull request by ScXfjiang**
Frontend

related files: `job_conf.proto`,  `optimizer.py`, `function_util.py`, `job_build_and_infer_ctx.cpp`

主要改动：

* 修改job_conf.proto支持multi optimizer；
* 修改optimizer相关的API（所有改动均保持向后兼容）：
  * 增加CombinedOptimizer，用户需使用该接口配置multi optimizer；
  * 为每个optimizer增加`variables`参数指定需要维护的权值，默认值为job内全部variables；

Backend

related files: `auto_learning_rate.cpp`, `generate_backward_and_optimizer_op_confs.cpp`, `optimizer.h`,  `optimizer.cpp`, `sgd_optm.cpp`,  `momentum_optm.cpp`,  `adam_optm.cpp`, `lambda_optm.cpp`, `lars_optm.cpp`, `rmsprop_optm.cpp`, 

主要改动：

* 遍历train_conf中的optimizer_conf字段，向网络中添加不同类型的Optimizer Op，每一种Optimizer配置均添加Auto Learning Rate模块；
* clip_by_global_norm中global语义从整个网络修改为本Optimizer内；
* 修改创建OptimizerOpConf的接口，使其传入更少的参数；

Test

related files: `test_multi_optimizer.py`

设计文档：https://github.com/Oneflow-Inc/OneTeam/issues/274

### CODE REVIEW

File: oneflow/core/job/job_conf.proto [hunk 1]
```diff
@@ -72,11 +72,14 @@ message WeightDecayConf {
...
 }
 
-message NormalModelUpdateOpUserConf {
-  optional LearningRateDecayConf learning_rate_decay = 1;
-  optional WarmupConf warmup_conf = 2;
-  optional ClipConf clip_conf = 3;
-  optional WeightDecayConf weight_decay_conf = 4;
+message OptimizerConf {
+  repeated string variables = 1;
+  optional float base_learning_rate = 2;
```
[the remark this answers is not in the document]
Comment (ScXfjiang) [reply oneflow/core/job/job_conf.proto:77]: 这里PyTorch使用base_learning_rate，TensorFlow使用initial_learning_rate。我更倾向与base_learning_rate，LrScheduler作为base_learning_rate的函数，在训练过程中不需要修改这个值。

**Review by strint — APPROVED** [review]
look good to me

File: oneflow/python/ops/optimizer.py [hunk 2]
```diff
@@ -1086,21 +1089,31 @@ def __init__(
...
+            self.variables = list(variables())
+        else:
+            self.variables = list(variables)
 
-    def _SetSpecificFieldsInTrainConf(self, train_conf):
+    def _AddOptimizerConfInTrainConf(self, train_conf) -> None:
+        optimizer_conf = train_conf.optimizer_conf.add()
+        self.lr_scheduler.SetLrFieldsInOptimizerConf(optimizer_conf)
+        if self.grad_clipping is not None:
+            optimizer_conf.clip_conf.CopyFrom(self.grad_clipping.clip_conf)
```
Comment (daquexian) [inline oneflow/python/ops/optimizer.py:1111]: 这个 grad_clipping 之前是基类的成员，现在为什么放到每一个子类里了呢 0。0

File: oneflow/python/ops/optimizer.py [hunk 3]
```diff
@@ -982,36 +984,34 @@ def SetLossScaleFieldsInTrainConf(self, train_conf):
...
             assert loss_scale_policy is None
             self.loss_scale_policy = StaticLossScalePolicy(loss_scale_factor)
         else:
             self.loss_scale_policy = loss_scale_policy
-        self.lr_scheduler = lr_scheduler
-        self.grad_clipping = grad_clipping
-        self.train_step_lbn = train_step_lbn
 
-    def _SetSpecificFieldsInTrainConf(self, train_conf):
+    def Variables(self) -> Sequence[Text]:
```
Comment (daquexian) [inline oneflow/python/ops/optimizer.py:998]: 返回值类型可以直接写 List[Text]

File: oneflow/python/ops/optimizer.py (hunk 2 above)
Comment (ScXfjiang) [reply oneflow/python/ops/optimizer.py:1111]: 这是因为现在CombinedOptimizer和其他Optimizer都是Optimizer子类，它们处理grad_clipping和lr_scheduler的行为不同，所以把它们放到子类。目前Optimizer基类仅处理整个Job内唯一的逻辑，例如train_step、loss scale。

File: oneflow/python/ops/optimizer.py (hunk 3 above)
Comment (ScXfjiang) [reply oneflow/python/ops/optimizer.py]: 已修改

File: oneflow/python/ops/optimizer.py (hunk 2 above)
Comment (daquexian) [reply oneflow/python/ops/optimizer.py:1111]: 哦哦，我觉得是不是再创建一个叫 BaseOptimizer（或者其它名字）的类作为除了 CombinedOptimizer 之外的 optimizer 的基类，来处理 grad_clipping 这样共同的逻辑，否则这段逻辑要在多个类里重复

Comment (ScXfjiang) [reply oneflow/python/ops/optimizer.py:1111]: 目前BaseOptimizer（先不考虑命名问题）可以抽出的公共逻辑较少，只有grad_clipping和lr_scheduler，代码量也很少。我更倾向于保留现在的设计保持Optimizer继承体系的简单性，等我们有更充分的理由时再考虑扩展Optimizer继承体系。

Comment (ScXfjiang) [reply oneflow/python/ops/optimizer.py:1111]: 这里应该达成一致了，暂时不对Optimizer继承体系做修改～
